### ISSUE TITLE
Contradictory BCEs in classification loss

### ISSUE THREAD

**Issue by kate-feingold** [ultralytics/yolov5#3812]
## ❔Question
I found behavior in the loss that seems illogical and I'd like to validate if my understanding is correct.

Classification loss gets one-hot encoded labels
https://github.com/ultralytics/yolov5/blob/02719dde52a99f28603be383334028a7ab9f1e06/utils/loss.py#L146
while it's likely undesirable for some datatsets.

## Additional context

The loss here
https://github.com/ultralytics/yolov5/blob/02719dde52a99f28603be383334028a7ab9f1e06/utils/loss.py#L147
averages individual BCEs along XY axes of ```ps``` (```ps.shape[0] * ps.shape[1]``` BCEs in total), where ```ps[i, j]``` stores a probability before sigmoid produced by some anchor in a particular grid cell number ```i``` for a class ```j```, so classification is learnt in a multi-label fashion. 

The problem is that anchors among ```ps``` are not unique by design, cause it's not each and every anchor once, but instead anchors from all (anchor, target) pairs built here:
https://github.com/ultralytics/yolov5/blob/02719dde52a99f28603be383334028a7ab9f1e06/utils/loss.py#L118
Since **a few targets** can be assigned to **one anchor** on a grid, labels shouldn't be one-hot encoded. Anchors in ```ps``` should be unique, targets in ```t``` should store 1s for classes of targets assigned to it, 0s everywhere else.

Rights now one prediction (anchor in a grid cell predicting a class) can get:
* 0 and 1 as a label at the same time;
* 1 as a label twice,

(2 BCEs added in both cases).

Based on my checks on COCO 2017 there can be around 10% repeating anchors in ```ps``` each iteration.

**Comment by glenn-jocher** [comment]
@yurkovak nice investigative work. Everything you said is essentially correct, except that the losses are not averaged over height-width but rather are averaged over just just 1 dimension, the group of target-anchor matches, which are sparsely sampled (non-uniquely as you state) from the 3D anchor-height-width space.

This is all good and correct, I designed YOLOv5 with this loss function implementation intentionally as we observed improved empirical results on COCO using it.

There should be significantly more than 10% anchor overlap though in nearly all situations.

In any case, ML is an empirical science, so if you think you have a good idea for an update the best way to evaluate your idea is simply to create a branch, implement your idea, and train two models side by side, the default and your proposed idea.
